### ISSUE TITLE
[3.7.31][C][U2D2][XH540-W270-R] Issue : Need clarifications on the purpose of having a do-while loop in the read_write.c example

### ISSUE THREAD

**Issue by robotiscustomer** [ROBOTIS-GIT/DynamixelSDK#432]
ISSUE TEMPLATE ver. 1.1.1

BE CAREFUL!! FOLLOW THE RULES AS FOLLOWS, OR YOUR ISSUE WILL BE WON'T FIX ANYWAY

STEP 1. Check what you are suffering from :

I'm using DynamixelSDK ver. 3.7.31

I'm using C Language

I'm using U2D2 serial converter

I'm using XH540W270-R

and I'm having an issue

US Support asked me to post here. Need more clarification on the use of the do-while loop in the read_write.c example. 

STEP 2. Write Title as [3.7.31][C][U2D2][XH540-W270-R] Issue : Need clarifications on the purpose of having a do-while loop in the read_write.c example.

STEP 3. Delete all written here, and describe what your problem is

In the ~/Desktop/DynamixelSDK/c/example/protocol2.0/read_write/read_write.c example, I see a command is launched to tell the dynamixel to move to a target position. Then, there is a do-while loop that keeps printing out the position obtained from the encoder while the dynamixel is moving to the target position until the difference between the target and present positions are within a threshold. 

Q1: What is the purpose of using the loop? 

Q2: 
a) What is the difference between using the code given in the read_write.c example (shown below) and the approach of just executing the included code only once by removing the "do {" and "} while ... _THRESHOLD));" lines?

b) If I want to record the position of the dynamixel while the robot is moving for data logging, what is the best way to do it? Execute the while-do loop as is in the example and only save the last present position before the loop ends? Execute the "Read present position" command only once and save the data to a file and then ask the dynamixel to move to the next target position? Other approach open for your suggestion.

Q3: How do I determine the best DXL_MOVING_STATUS_THRESHOLD value to use? It seems that the smaller the value, the more times the do-while loop is executed and the closer the final present position is to the target position. However, due to inaccuracy of the encoder and other reasons, the loop may not terminate or take a long time to terminate. 

```C
do
    {
      // Read present position
      dxl_present_position = read2ByteTxRx(port_num, PROTOCOL_VERSION, DXL_ID, ADDR_MX_PRESENT_POSITION);
      if ((dxl_comm_result = getLastTxRxResult(port_num, PROTOCOL_VERSION)) != COMM_SUCCESS)
      {
        printTxRxResult(PROTOCOL_VERSION, dxl_comm_result);
      }
      else if ((dxl_error = getLastRxPacketError(port_num, PROTOCOL_VERSION)) != 0)
      {
        printRxPacketError(PROTOCOL_VERSION, dxl_error);
      }

      printf("[ID:%03d] GoalPos:%03d  PresPos:%03d\n", DXL_ID, dxl_goal_position[index], dxl_present_position);

    } while ((abs(dxl_goal_position[index] - dxl_present_position) > DXL_MOVING_STATUS_THRESHOLD));

```

**Comment by ROBOTIS-zerom** [comment]
A1 : As you know, it is to read and print the present position value at least once.

A2. a) I don't know the intention of the question.  the do-while loop is just to print the present position while moving to the goal position. If the do-while loop is removed, the present position value at the start of movement will printed once and no more changing present position values will be printed.

A2. b) It seems to depend on what data you want. 
In the example, the present position value and the goal position value are directly compared under the condition of the end of the loop, but the same effect can be obtained by using the moving value of the control table.
https://emanual.robotis.com/docs/en/dxl/p/ph54-200-s500-r/#moving 

A3. The suggested threshold value is for reference only, and you should find the optimal value yourself. 
As you explained, you may want to move more accurately even if it takes a long time, or you may want to move on to the next move when you get close, even if it is less accurate.

**Comment by robotiscustomer** [comment]
Thank you
